### ISSUE TITLE
Support for a mask during autoregressive generation with Key-Value Caching

### ISSUE THREAD

**Issue by Oufattole** [lucidrains/x-transformers#292]
Why isn't a mask supported when key-value caching is enabled [here](https://github.com/lucidrains/x-transformers/blob/abeedc8cb60180892f1c19b42548c9833abe4f5b/x_transformers/x_transformers.py#L1946)?

**Comment by lucidrains** [comment]
@Oufattole in what scenario would you need masking when doing autoregressive decoding?

**Comment by Oufattole** [comment]
I'm trying to do sliding window inference, but the lengths of the initial prompts are different in my transformer, so I think I should mask out the padding as that's what we do during autoregressive pretraining.

**Comment by lucidrains** [comment]
@Oufattole yes I see, so you are off the beaten path

sliding windows transformers isn't supported here yet

**Comment by Oufattole** [comment]
Ahhh I see thank you, I'll try that! With medical data, unlike in NLP and CV, many patient trajectories are very small and you don't need a long sequence length at all. For example, with my dataset 80% of patients are below the 512 max sequence length, but a small subset of patients are punching over 30k (this is after extreme reductions in the vocabulary -- i.e. which time-series variables we model, prior to which some of these patients hit over 300k).

I naively am trying to use sliding windows, but if there is a better approach you recommend for handling such extreme sequence length variations, I would be happy to try it.

**Comment by Oufattole** [comment]
Wait, actually, I think you do support masking the left padded tokens with the [seq_start_pos arg here](https://github.com/lucidrains/x-transformers/blob/abeedc8cb60180892f1c19b42548c9833abe4f5b/x_transformers/x_transformers.py#L1921C19-L1921C32) @lucidrains .

**Comment by lucidrains** [comment]
@Oufattole so that hyperparameter was actually built for variable prompt lengths iirc

during sliding window, you'll have to slice the cached key values as you decode out of the window length

**Comment by lucidrains** [comment]
@Oufattole what specialty is this and what exactly are you trending in the EMR that hits 300k in length?

**Comment by Oufattole** [comment]
Yes, I think you already do slice this kv-cache slicing during generation [here](https://github.com/lucidrains/x-transformers/blob/abeedc8cb60180892f1c19b42548c9833abe4f5b/x_transformers/autoregressive_wrapper.py#L230) when restricting to the max_seq_length (i.e. in the sliding window setting). Am I correct about this?

I'll send you an email in regard to the broader EHR modeling question, which I realize may be out of scope for this github issue.
